modules/rgbd/src/tsdf.cpp

```diff
@@ -388,27 +420,34 @@ struct IntegrateInvoker : ParallelLoopBody
                     Point2f projected = proj(camSpacePt, camPixVec);
 
                     depthType v = bilinearDepth(depth, projected);
-                    if(v == 0)
+                    if (v == 0) {
                         continue;
-
+                    }
+                    if (!pixNorm_flag && z==endZ-1)
+                        {
+                            pixNorm_o = (pixNorm - pixNorm_s) / (z - startZ);
+                            pixNorm_flag = true;
+                        }
                     // norm(camPixVec) produces double which is too slow
-                    float pixNorm = sqrt(camPixVec.dot(camPixVec));
+                    if (pixNorm_flag)
+                        pixNorm += pixNorm_o;
```

Comment:
The fact that this approximation works is not more than a happy coincidence. Changing a pose more radically would fail this. Here's how `pixNorm` function looks: https://www.geogebra.org/3d/pretpsdt where the typical x, y are between -1 and +1, i.e. in the area of maximal non-linearity. Imagine a straight line projected from the Oxy plane onto that figure: it would be far from a straight line.
The optimization I was talking about is pre-calculation of `pixNorm` for all pixels and accessing it instead of calculation.